### PR TITLE
Add SpatialDerivativeLoss to physicsnemo.experimental.losses

### LINKED ISSUE #1805: Add a generic spatial-derivative regularization loss to physicsnemo.experimental.losses
### Motivation

Neural-operator surrogates on structured grids often benefit from a
regularizer that matches the spatial *gradients* of the prediction and target,
not just their values. PhysicsNeMo has grid derivative operators in
`physicsnemo.nn.functional.derivatives`, but they are periodic, coordinate-based,
and operate on a single scalar field — they do not cover a batched, non-periodic,
mask-aware derivative-comparison **loss** (needed for sparse grids).

### Proposed change

Add `SpatialDerivativeLoss` to a new `physicsnemo.experimental.losses` package:
a generic, dimension-agnostic loss that compares interior (non-periodic)
central-difference gradients of `pred` vs `target` on a non-uniform grid
(explicit per-axis cell widths), with a dependency-injected data-fitting metric
callable (e.g. from `physicsnemo.metrics.general`) and optional inactive-cell
masking. Reusable `central_difference` / `cell_centre_distance` helpers are
exported. Includes unit tests and a CHANGELOG entry.

### Scope

New experimental API. Part of upstreaming reusable components from the Neural
Operator Factory example. Depends on the generic data-fitting metrics.

### PR DESCRIPTION
**Pull request by wdyab**
## Summary

Adds a new experimental package `physicsnemo.experimental.losses` with
`SpatialDerivativeLoss`: a generic, dimension-agnostic spatial-derivative
regularization loss. For each requested spatial axis it compares the interior
(non-periodic) central-difference gradient of `pred` and `target` on a
non-uniform grid (explicit per-axis cell widths) using a dependency-injected
data-fitting `metric` callable (e.g. any `physicsnemo.metrics.general` function),
and supports an optional inactive-cell mask for sparse grids. The reusable
`central_difference` / `cell_centre_distance` helpers are also exported.

**No duplication:** the existing `physicsnemo.nn.functional.derivatives`
operators (`rectilinear_grid_gradient`, `uniform_grid_gradient`, and the newer
`*_divergence`/`*_curl`/`*_laplacian`) are periodic, coordinate-based, and
single-field, so they do not cover this batched, non-periodic, mask-aware
derivative-comparison *loss*. This rationale is documented in the module.

Closes #1805

## Test plan

- [x] `test/experimental/losses/test_derivative_loss.py`: FD-helper correctness
      (uniform + non-uniform grids), 2D/3D, masking (boundary exclusion,
      per-sample union, sparse-grid no-NaN), gradient flow, dependency-injected
      metric sweep, minimum-grid and error-path cases.
- [x] `interrogate` 100%, `ruff` check+format clean, license headers, CHANGELOG.
- [x] Doctest example in the class docstring passes.

### CODE REVIEW

File: physicsnemo/experimental/losses/derivative.py [hunk 1]
```diff
@@ -0,0 +1,246 @@
...
+                t = d_target
+
+            # Skip axes with no active derivative samples (e.g. a fully-masked
+            # region) so that reducing an empty tensor cannot produce NaN.
+            if p.numel() == 0:
+                continue
+
+            total = total + self.metric(p, t)
+
+        return total / len(cell_widths)
```
[the remark this answers is not in the document]
Comment (wdyab) [reply physicsnemo/experimental/losses/derivative.py]: Fixed in 4a98af41: the loss now divides by the number of *contributing* axes (those with at least one valid derivative sample) and returns 0 when none contribute, so a fully-masked axis no longer silently shrinks the loss magnitude. Documented the contract in the Outputs section and added `test_masked_axis_excluded_from_average` + `test_fully_masked_returns_zero`.

File: physicsnemo/experimental/losses/derivative.py [hunk 2]
```diff
@@ -0,0 +1,246 @@
...
+                p = d_pred[sel]
+                t = d_target[sel]
+            else:
+                p = d_pred
+                t = d_target
+
+            # Skip axes with no active derivative samples (e.g. a fully-masked
+            # region) so that reducing an empty tensor cannot produce NaN.
+            if p.numel() == 0:
+                continue
```
[the remark this answers is not in the document]
Comment (wdyab) [reply physicsnemo/experimental/losses/derivative.py]: Documented rather than rewritten. The masked path intentionally selects active samples and computes a mean over *active* cells; the suggested masked-fill + full-tensor reduction would change the loss value (mean over all cells vs. active cells) and diverge from the masked-mean semantics this reproduces. Added a Notes section stating the masked path uses data-dependent shapes and is therefore not `torch.compile`-traceable, while the unmasked path compiles normally. The `is_compiling()` guard is standard shape-validation skipping, not a compile guarantee.

File: physicsnemo/experimental/losses/derivative.py [hunk 3]
```diff
@@ -0,0 +1,246 @@
...
+            # Zero inactive cells so the stencil does not read garbage across the
+            # active/inactive boundary. ``(*spatial) -> (1, *spatial, 1)``.
+            mask_exp = union_mask.unsqueeze(0).unsqueeze(-1)
+            pred = pred * mask_exp
+            target = target * mask_exp
+
+        total = torch.zeros((), device=pred.device, dtype=pred.dtype)
+        for axis, widths in cell_widths.items():
+            spacing = cell_centre_distance(widths, min_spacing=self.min_spacing)
+            d_pred = central_difference(pred, axis, spacing)
```
[the remark this answers is not in the document]
Comment (wdyab) [reply physicsnemo/experimental/losses/derivative.py:253]: Fixed in 4a98af41: `forward` now validates that every `cell_widths` key is a spatial axis in [1, ndim-2] (the batch axis 0 and the trailing feature/time axis are excluded), under the `not torch.compiler.is_compiling()` guard. Added parametrized `test_invalid_axis_key_raises` for axis 0 and the feature axis.

**Comment by wdyab** [comment]
Follow-up: added a guard (under the `not torch.compiler.is_compiling()` block) that each `cell_widths[axis]` is a 1-D tensor of length `pred.shape[axis]`, so a widths/axis-size mismatch fails fast with a clear message instead of a downstream broadcast error. Added `test_widths_length_mismatch_raises`. (Addresses the widths-length note from the last review.)

**Comment by wdyab** [comment]
@melo-gonzo for your review. I closed the other PR based on your recommendation

**Comment by wdyab** [comment]
@mnabian  @peterdsharpe for your review; to be added to metrics

File: physicsnemo/experimental/losses/derivative.py [hunk 4]
```diff
@@ -0,0 +1,289 @@
...
+"""
+
+from typing import Callable, Dict, Optional
+
+import torch
+import torch.nn as nn
+
+from physicsnemo.metrics.general.mse import mse
+
+Tensor = torch.Tensor
```
Comment (peterdsharpe) [inline physicsnemo/experimental/losses/derivative.py:44]: Why?

File: physicsnemo/experimental/losses/derivative.py [hunk 5]
```diff
@@ -0,0 +1,289 @@
...
+
+
+def cell_centre_distance(cell_widths: Tensor, min_spacing: float = 1e-6) -> Tensor:
```
Comment (peterdsharpe) [inline physicsnemo/experimental/losses/derivative.py:47]: Repo rules require jaxtyping

File: physicsnemo/experimental/losses/derivative.py [hunk 6]
```diff
@@ -0,0 +1,289 @@
...
+    ``uniform_grid_gradient``). Those operators are *periodic*, coordinate-based,
+    and act on a single scalar field, which makes them unsuitable for this loss:
+    the regularizer must operate on *batched* fields with a trailing feature/time
+    axis, use *non-periodic* interior central differences (so boundary cells are
+    not wrapped), and support an inactive-cell *mask* (sparse grids such as the
+    Norne reservoir). The self-contained :func:`central_difference` /
+    :func:`cell_centre_distance` helpers below cover exactly that regime.
+"""
+
+from typing import Callable, Dict, Optional
```
Comment (peterdsharpe) [inline physicsnemo/experimental/losses/derivative.py:37]: Prefer `dict`, `| None`

**Review by peterdsharpe — CHANGES_REQUESTED** [review]
This is part of the already-closed #1807, right? Let's find a time to discuss the high-level value prop and API design offline prior to opening new PRs. Thanks!

Our starting point for discussion aligns with agent review of this PR, but open to being convinced:

> I'd not merge this as-is. This is a good-faith PR with clean code and passing tests (I ran them: 59 passed), but strategically it's a parallel, numerically-inferior reimplementation of a derivative operator you already have, wrapped in a loss whose novel parts are specific to the author's one example. Nothing here is blocked by not merging - they can keep all of this in their own recipe downstream. The bar you articulated ("applicability beyond the specific case they're working with") is not met by the current design.

**Comment by wdyab** [comment]
Closing per the offline discussion with the maintainers: these regularization losses will live in the reservoir-simulation example recipe downstream, and the generic finite-difference need will be addressed by extending physicsnemo.nn.functional.derivatives (non-periodic + batched) rather than adding a parallel implementation. Thanks for the reviews!
